File: treee/230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> vec;
        inorder(root, vec, k);
        return *vec.rbegin();
    }
    
    void inorder(TreeNode* root, vector<int>& vec, int k) {
        if (root == nullptr)
            return;
        if (vec.size() == k)
            return;
        
        inorder(root->left, vec, k);
        if (vec.size() < k)
            vec.push_back(root->val);
        inorder(root->right, vec, k);
    }
};

File: treee/230-kth-smallest-element-in-a-bst.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    import scala.collection.mutable.ListBuffer
    def kthSmallest(root: TreeNode, k: Int): Int = {
        def inorder(root: TreeNode, l: ListBuffer[Int], k: Int): ListBuffer[Int] = {
            if (root == null || l.size == k) {
                l
            } else {
                var r = inorder(root.left, l, k)
                if (r.size < k) {
                    r = r :+ root.value
                }
                inorder(root.right, r, k)
            }
        }

        val r = inorder(root, ListBuffer(), k)
        r.last
    }
}
